normative/dtpr-stickers
# Feature/migration

### Notion card link(s)
- [Migrate Realtime DB > Firestore DB](https://www.notion.so/normative/Migrate-Realtime-DB-Firestore-DB-2bb5a87c4c864a17969aef211ae4f23f)
- [Add "customer" table to firestore instances](https://www.notion.so/normative/Add-customer-table-to-firestore-instances-9a37e1b0459848b680f0f04aa06f0a32)

### Description (Proposed changes, additional context, chosen approach/alternative approaches, gotchas, etc)
- Legacy JSON data from realtime database is manually exported and saved in a new migration/data folder (and added to `.gitignore`)
- `migration/scripts/migration.js` is a node script that migrates one legacy JSON data file to one of the new firestore instances at a time.
- Currently [Boston Pilot legacy data](https://console.firebase.google.com/project/dtpr-boston/database/dtpr-boston-default-rtdb/data) has been migrated to dtpr-usa firestore instance, and shared legacy data (Boston Pilot & Normative - copied from [Kei's slack thread](https://normative.slack.com/archives/C01G91SGX6U/p1632170967036900?thread_ts=1632170658.035500&cid=C01G91SGX6U)) has been migrated to dtpr-canada instance. 
- We have not yet culled data unless it has been moved elsewhere (e.g. datachain & children have been moved to new datachain subcollection of sensors).
- Except for "users" object in shared legacy data; this has not been migrated as we aren't clear if/how we'll be using this data yet.
- In shared legacy data I had to manually update `sensors/-MSTrI2eY2KgpnGr4E4f/datachain/-MV7Y9UTmoaFUQ6cKGQ4` to an object so that it would migrate: 
```
"-MV7Y9UTmoaFUQ6cKGQ4": {
          "additionalDescription": "Test",
          "priority": 0,
          "icon" : "purpose/enforcement",
          "category" : "purpose",
          "title" : "Enforcement",
          "headline": "Purpose of this technology",
          "description": "Used for enforcement of rules or regulations. Find out more here"
        },
```

### Steps for reviewer(s) to reproduce/test
The following has also been added to the docs as a part of this pull request:
- Manually export legacy data as JSON & save in new data folder: 
  `root/migration/data/{LEGACY DATA}`.
- Copy the target firestore instance's service account key & save as 
  `env.production-{FIRESTORE INSTANCE}` in `migration/scripts` folder.
- IMPORTANT: Update `migration/.gitignore` to include all legacy data & service account information.
- In `migration/migation.js`
    1. Add/update service account variable:
       `const serviceAccountUSA = require('./env.production-{FIRESTORE INSTANCE}')`;
    2. Update admin.initializeApp credential to your chosen firestore instance:
       `credential: admin.credential.cert({SERVICE ACCOUNT VARIABLE}),`
    3. Add/update data location variable:
       `const legacyData = require('../data/legacyData-{LEGACY DATA}.json')`
- In your terminal run `node migration/scripts/migration.js` to migrate legacy data to chosen firestore instance.
- If you see `TypeError: Cannot convert undefined or null to object` in terminal, check legacy data is valid JSON.
